### PR TITLE
Fix Issues in MCC Metrics and Loss Calculations

### PR DESCRIPTION
**Pull request by shihchengli**
## Description
Address #846 and #940. 

Two major adjustments in this PR include:
1. Providing task and data weights as arrays of ones to avoid `None` being used in metrics calculations.
2. Per the comment in #846, `MccMixin` can be removed and a small value, 1e-8, is added to the denominator to avoid the loss being calculated as NaN.

## Relevant issues
#846 #940 

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/nn/loss.py [hunk 1]
```diff
@@ -182,41 +181,48 @@ def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, *args) -> Tensor:
...
-    .. [mccSklearn] https://scikit-learn.org/stable/modules/generated/sklearn.metrics.matthews_corrcoef.html
-    """
-
-    def __call__(self, preds: Tensor, targets: Tensor, mask: Tensor, weights: Tensor, *args):
+@LossFunctionRegistry.register("binary-mcc")
+class BinaryMCCLoss(LossFunction):
+    def forward(self, preds: Tensor, targets: Tensor, mask: Tensor, weights: Tensor, *args):
         if not (0 <= preds.min() and preds.max() <= 1):  # assume logits
-            preds = preds.softmax(2)
+            preds = preds.sigmoid()
```
Comment (hwpang) [inline chemprop/nn/loss.py:188]: Why do we change this here?

File: chemprop/nn/loss.py [hunk 2]
```diff
@@ -182,41 +181,48 @@ def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, *args) -> Tensor:
...
+
+    References
+    ----------
+    .. [mccWiki] https://en.wikipedia.org/wiki/Phi_coefficient#Multiclass_case
+    .. [mccSklearn] https://scikit-learn.org/stable/modules/generated/sklearn.metrics.matthews_corrcoef.html
+    """
+
+    def forward(self, preds: Tensor, targets: Tensor, mask: Tensor, weights: Tensor, *args):
+        if not (0 <= preds.min() and preds.max() <= 1):  # assume logits
+            preds = preds.softmax(2)
```
Comment (hwpang) [inline chemprop/nn/loss.py:219]: Why do we use softmax here but sigmoid in the metric?

File: chemprop/nn/predictors.py [hunk 3]
```diff
@@ -244,7 +244,7 @@ def forward(self, Z: Tensor) -> Tensor:
     def train_step(self, Z: Tensor) -> Tensor:
         Y = super().forward(Z)
 
-        F.softplus(Y) + 1
+        return F.softplus(Y) + 1
```
Comment (hwpang) [inline chemprop/nn/predictors.py:247]: Interesting that we didn't catch this earlier. This reminds me that we need to add a test for dirichlet. It doesn't need to be added here in this PR.

File: chemprop/cli/train.py [hunk 4]
```diff
@@ -875,6 +876,7 @@ def train_model(
             targets = test_dset.Y
             mask = torch.from_numpy(np.isfinite(targets))
             targets = np.nan_to_num(targets, nan=0.0)
+            weights = torch.ones(len(test_dset))
```
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/cli/train.py:879]: This is per-sample weight

File: chemprop/nn/loss.py (hunk 1 above)
Comment (shihchengli) [reply chemprop/nn/loss.py:188]: Because this is for binary classification, the activation function should be sigmoid. This bug was identified in #846.

File: chemprop/nn/loss.py (hunk 2 above)
Comment (shihchengli) [reply chemprop/nn/loss.py:219]: The reason to keep softmax here is that this is a multiclass classification.

File: chemprop/nn/predictors.py (hunk 3 above)
Comment (shihchengli) [reply chemprop/nn/predictors.py:247]: It is probably because we haven't extensively tested the more advanced model settings yet, so no one found this bug. Also, I would be happy to include more comprehensive integration tests after this PR.

**Review by hwpang — APPROVED** [review]
LGTM
